python/services/evaluationservice/dmod/evaluationservice/evaluation_service/consumers/action.py

```diff
@@ -0,0 +1,939 @@
+
+    return function_with_parameters
+
+
+class ActionDescriber:
+    @classmethod
+    def get_client_name(cls):
+        consumer_pattern = re.compile("[Cc][Oo][Nn][Ss][Uu][Mm][Ee][Rr]")
+        class_name = consumer_pattern.sub("", cls.__name__) + "Client"
+        return class_name
+
+    @classmethod
+    def get_action_handlers(cls) -> typing.Dict[str, typing.Callable[[typing.Dict[str, typing.Any]], typing.Coroutine]]:
+        """
+        Generates a mapping of names for socket actions to functions that bear a dictionary of their required
+        parameters mapped to human friendly strings of their expected data types
+
+        Returns:
+            A dictionary mapping the name of actions a client may request to the functions that handle them
+        """
+        def callable_requires_parameters(member) -> bool:
+            """
+            Checks to see if a passed member counts as an action handler
+
+            Args:
+                member: A member retrieved by calling `inspect.get_members`
+
+            Returns:
+                Whether the passed member is a callable with a  `required_parameters` map
+            """
+            return isinstance(member, typing.Callable) \
+                   and hasattr(member, 'required_parameters') \
+                   and isinstance(getattr(member, 'required_parameters'), dict)
+
+        return {
+            handler_name: handler
+            for handler_name, handler in inspect.getmembers(cls, predicate=callable_requires_parameters)
+        }
+
+    @classmethod
+    def generate_action_catalog(cls) -> typing.Sequence[typing.Mapping[str, typing.Union[str, typing.Mapping[str, typing.Any]]]]:
+        actions = list()
+
+        for name, function in cls.get_action_handlers().items():
+            # `function` will be a function like `(dict) => Coroutine`, where the function will have a dictionary
+            # mapping the name of required parameters for each handler to a human friendly description of its type
+            documentation = ""
+
+            if function.__doc__:
+                split_documentation = function.__doc__.strip().split("\n")
+                if len(split_documentation) > 0:
+                    documentation = split_documentation[0]
+
+            descriptor = {
+                "action": name,
+                "documentation": documentation
+            }
+
+            action_parameters = {
+                parameter_name: {
+                    "type": parameter_type,
+                    "required": False
+                }
+                for parameter_name, parameter_type in getattr(function, "optional_parameters", dict()).items()
+            }
+
+            action_parameters.update({
+                parameter_name: {
+                    "type": parameter_type,
+                    "required": True
+                }
+                for parameter_name, parameter_type in getattr(function, "required_parameters").items()
```

Comment:
It's safe here because `get_action_handlers` defines an action handler as a callable that bears a "required_parameters" dictionary. If it doesn't bear the decorator, it isn't considered an action handler. Now, if it doesn't have anything within that dictionary, nothing is added in this section. `LaunchConsumer.get_actions` and `LaunchConsumer.generate_library`, for example, have the decorator but no parameters set. It might be more appropriate, now that the system has evolved a bit, to rename the decorator to something along the lines of `action_handler` with fields for 'required_parameters' and 'optional_parameters'. Do you think I should add a ticket?